build.rs

```diff
@@ -64,159 +64,488 @@
+        asm_args.push(String::from("crypto/libring-asm.Windows.vcxproj"));
+        run_command_with_args(&msbuild, &asm_args);
+    } else {
+        // TODO: rename msvc-ring-asm-i686.lib to msvc-ring-asm-x86.lib
+        let pregenerated_lib_name = format!("msvc-ring-asm-{}.lib",
+                                            target_info.target_arch());
+        let mut pregenerated_lib = PathBuf::from("pregenerated");
+        pregenerated_lib.push(pregenerated_lib_name);
 
-        let mut core_args = args.clone();
-        core_args.push(String::from("crypto/libring.Windows.vcxproj"));
-        run_command_with_args(&msbuild, &core_args);
+        let ring_asm_lib_name = String::from("ring-asm.lib");
+        let mut ring_asm_lib = lib_path.clone();
+        ring_asm_lib.push(&ring_asm_lib_name);
+        println!("{:?} -> {:?}", &pregenerated_lib, &ring_asm_lib);
 
-        let mut test_args = args.clone();
-        test_args.push(String::from("crypto/libring-test.Windows.vcxproj"));
-        run_command_with_args(&msbuild, &test_args);
+        std::fs::create_dir_all(&lib_path).unwrap();
+        let _ = std::fs::copy(&pregenerated_lib, &ring_asm_lib).unwrap();
     }
+    println!("cargo:rustc-link-lib=static=ring-asm");
+
+    let mut core_args = args.clone();
+    core_args.push(String::from("crypto/libring.Windows.vcxproj"));
+    run_command_with_args(&msbuild, &core_args);
 
-    println!("cargo:rustc-link-search=native={}", lib_path.to_str().unwrap());
-    println!("cargo:rustc-link-lib=static={}-core", LIB_NAME);
+    let mut test_args = args.clone();
+    test_args.push(String::from("crypto/libring-test.Windows.vcxproj"));
+    run_command_with_args(&msbuild, &test_args);
 
+    println!("cargo:rustc-link-lib=static=ring-core");
     // XXX: Ideally, this would only happen for `cargo test`, but we don't know
     // how to do that yet.
-    println!("cargo:rustc-link-lib=static={}-test", LIB_NAME);
-    if !use_msbuild {
-        println!("cargo:rustc-flags=-l dylib=stdc++");
-    }
-
-    Ok(())
+    println!("cargo:rustc-link-lib=static=ring-test");
 }
 
 fn run_command_with_args<S>(command_name: S, args: &Vec<String>)
-                            where S: AsRef<std::ffi::OsStr> + Copy {
+    where S: AsRef<std::ffi::OsStr> + Copy
+{
     if !std::process::Command::new(command_name)
-            .args(&args)
-            .status()
-            .unwrap_or_else(|e| { panic!("failed to execute {}: {}",
-                            command_name.as_ref().to_str().unwrap(), e); })
-            .success() {
+        .args(&args)
+        .status()
+        .unwrap_or_else(|e| {
+            panic!("failed to execute {}: {}",
+                   command_name.as_ref().to_str().unwrap(),
+                   e);
+        })
+        .success() {
         panic!("{} execution failed", command_name.as_ref().to_str().unwrap());
     }
 }
+
+fn compile(file: &str, target_info: &TargetInfo, mut out_dir: PathBuf)
```

Comment:
The gcc crate is not capable to distinguish between different kind of files. If I pass all RING_TEST_SRCS to the gcc crate I need to set the [cpp flag](https://docs.rs/gcc/0.3.38/gcc/struct.Config.html#method.cpp). This means everything is compiled as c++ file. (The gcc crate simply calls CC or CXX on all passed files.) At this point some of the supported compilers are beginning to complain about incorrect compiler and linker flags. For example see this [build](https://travis-ci.org/briansmith/ring/jobs/170415102).
Because of this I decided to compile every file on it's own. It should be simple to call this in parallel by using par_iter form [rayon](https://crates.io/crates/rayon) like the gcc crate.